src/servers/src/http/loki.rs

```diff
@@ -0,0 +1,387 @@
+        },
+        ColumnSchema {
+            column_name: LOKI_LINE_COLUMN.to_string(),
+            datatype: ColumnDataType::String.into(),
+            semantic_type: SemanticType::Field.into(),
+            datatype_extension: None,
+            options: None,
+        },
+    ];
+}
+
+#[axum_macros::debug_handler]
+pub async fn loki_ingest(
+    State(log_state): State<LogState>,
+    Extension(mut ctx): Extension<QueryContext>,
+    TypedHeader(content_type): TypedHeader<ContentType>,
+    LogTableName(table_name): LogTableName,
+    bytes: Bytes,
+) -> Result<HttpResponse> {
+    ctx.set_channel(Channel::Loki);
+    let ctx = Arc::new(ctx);
+    let table_name = table_name.unwrap_or_else(|| LOKI_TABLE_NAME.to_string());
+    let db = ctx.get_db_string();
+    let db_str = db.as_str();
+    let exec_timer = Instant::now();
+
+    // init schemas
+    let mut schemas = LOKI_INIT_SCHEMAS.clone();
+
+    let mut rows = match content_type {
+        x if x == *JSON_CONTENT_TYPE => handle_json_req(bytes, &mut schemas).await,
+        x if x == *PB_CONTENT_TYPE => handle_pb_req(bytes, &mut schemas).await,
+        _ => UnsupportedContentTypeSnafu { content_type }.fail(),
+    }?;
+
+    // fill Null for missing values
+    for row in rows.iter_mut() {
+        if row.len() < schemas.len() {
+            for _ in row.len()..schemas.len() {
+                row.push(GreptimeValue { value_data: None });
+            }
+        }
+    }
+
+    let rows = Rows {
+        rows: rows.into_iter().map(|values| Row { values }).collect(),
+        schema: schemas,
+    };
+    let ins_req = RowInsertRequest {
+        table_name,
+        rows: Some(rows),
+    };
+    let ins_reqs = RowInsertRequests {
+        inserts: vec![ins_req],
+    };
+
+    let handler = log_state.log_handler;
+    let output = handler.insert(ins_reqs, ctx).await;
+
+    if let Ok(Output {
+        data: OutputData::AffectedRows(rows),
+        meta: _,
+    }) = &output
+    {
+        METRIC_LOKI_LOGS_INGESTION_COUNTER
+            .with_label_values(&[db_str])
+            .inc_by(*rows as u64);
+        METRIC_LOKI_LOGS_INGESTION_ELAPSED
+            .with_label_values(&[db_str, METRIC_SUCCESS_VALUE])
+            .observe(exec_timer.elapsed().as_secs_f64());
+    } else {
+        METRIC_LOKI_LOGS_INGESTION_ELAPSED
+            .with_label_values(&[db_str, METRIC_FAILURE_VALUE])
+            .observe(exec_timer.elapsed().as_secs_f64());
+    }
+
+    let response = GreptimedbV1Response::from_output(vec![output])
+        .await
+        .with_execution_time(exec_timer.elapsed().as_millis() as u64);
+    Ok(response)
+}
+
+async fn handle_json_req(
+    bytes: Bytes,
+    schemas: &mut Vec<ColumnSchema>,
+) -> Result<Vec<Vec<GreptimeValue>>> {
+    let mut global_label_key_index: HashMap<String, u16> = HashMap::new();
+    global_label_key_index.insert(GREPTIME_TIMESTAMP.to_string(), 0);
+    global_label_key_index.insert(LOKI_LINE_COLUMN.to_string(), 1);
```

Comment:
Are we using line number as tag here?

Reply:
Renaming `global_label_key_index` to `column_indexer`, as it's just a record pair for the column name and its position. The line number is irrelevant here.